### DISCUSSION TITLE
How to properly run sbws in Shadow with tornettools?

### DISCUSSION THREAD

**Discussion by higher-order** [shadow/shadow#3306] [General]
There was a post 2 years ago about running [sbws](https://gitlab.torproject.org/tpo/network-health/sbws) (the bandwidth scanner) in Shadow: https://github.com/shadow/shadow/discussions/1909
However, back then it seemed to be an open question.

I'd like to ask if it's clear now regarding how to run sbws in Shadow?
Additionally, I'm also using tornettools, which generates shadow config for me.
Is it as simple as inserting:
```
- path: <path to sbws>
  args: ...
  ...
```
To the `processes` block of each 4uthority host in `shadow.config.yaml` which is generated by tornettools?
Or is it more complicated? To my understanding, without sbws, the dirauths in Shadow read the bandwidth files provided by Shadow. So it seems like at the very least I'll need to reconfigure them to read the bandwidth files from sbws.

**Comment by robgjansen** [comment]
Hello! Just a word of warning: I have never tried to run sbws myself, so this is general advice. You may have to play around and adjust things to get it to work.

sbws is just python code, and we have successfully run other python applications in Shadow before. So I believe it _should_ be possible to do what you want now. Whether or not it works successfully depends a bit on the python features sbws is using - if it uses any Linux features that Shadow does not support yet, then you might get some errors in the Shadow or sbws log files. But you could try it and see what happens!

As far as _how_ to run sbws in Shadow, you generally have the right idea. You need to configure a host in the Shadow config file to run sbws. You _could_ add it to the existing authority hosts by adding a new line into their `processes` block, but I think you actually want to add a completely new `bwauthority` host and add it to the new host's process block instead. That way you can specify a higher bandwidth capacity if needed, and the bandwidth measurements won't negatively affect the Tor directory authority duties. I think this also better reflects the way the live Tor network is configured.

You might need to specify `python` as the binary, and the top-level `sbws` script in the `args` field (along with any other args needed by sbws). This assume you have already installed `sbws` globally and it is available to run on the system. And you may need to set the PYTHONPATH environment variable in the shadow config too?

The network configuration already stores a `v3bw` file in a bandwidth authority path. For example, in our minimal example, the authority configures the bandwidth file location here:

https://github.com/shadow/shadow/blob/df2db4a760ece582dc8956c82a53555c83af24f6/src/test/tor/minimal/conf/tor.authority.torrc#L3

and that file is specified here:

https://github.com/shadow/shadow/blob/df2db4a760ece582dc8956c82a53555c83af24f6/src/test/tor/minimal/shadow.data.template/hosts/torflowauthority/v3bw#L1-L10

If using `tornettools`, then the location would be `hosts/bwauthority/v3bw`.

You probably want to leave this critical file in place to start with, so that the network has a valid v3bw file while sbws is conducting measurements. Then when the measurements are completed, produce a new v3bw file and overwrite the `v3bw` to replace it with your new version from sbws.

Let us know if you get this working!

**Reply by robgjansen** [reply]
Oh, I should have also mentioned that it can take many hours to complete a full set of measurements with sbws, depending on how many relays are in your network. However, `tornettools` simulations generally only run for 1 simulated hour (you can of course increase that setting in the shadow.config file, but it will take more real time to simulate more simulated hours). `tornettools` currently do not bother automatically adding an active measurement step (sbws, torflow, etc.) for this reason.

**Reply by higher-order** [reply]
Hi Rob, thank you very much for the hints. I like how a dirauth can directly read the v3bw file generated by the bwauth even if it's running on a different virtual host, because everything is simulated on the same physical host. I have since made quite some progress.

sbws requires a destination server which also has TLS enabled. I have tried both apache and nginx, but the former had some trouble with systemd, bus, and sudo, while the latter ran into some unexpected EOF when being requested with curl in Shadow. I eventually settled with a simple HTTPS server in Python I found here: https://gist.github.com/DannyHinshaw/a3ac5991d66a2fe6d97a569c6cdac534. For testing, sbws does allow self-signed certs.

I put the simple server into the 1% network generated by tornettools, and added another client that curls it with the `--insecure` flag. The client was able to download the 1G file and get a reply from `/postpath`, as required by sbws, in Shadow.

I've also tested sbws with the simple server on the real network for a few minutes. The test was very short to avoid congesting the Tor network. But sbws was able to scan some relays and generate meaningful relay lines in the v3bw file.

However, putting the same sbws and simple server into Shadow and the 1% network generated by tornettools caused some problems. The sbws exited with these errors:
```
Jan 01 00:05:10 sbws[1000]: <INFO> (MainThread) sbws 0+unknown with python 3.10.12 on shadowsys-shadowrelease-shadowmachine-64bit, stem 1.8.2, and requests 2.31.0
conf <configparser.ConfigParser object at 0x7ffff4dc5e70>
Jan 01 00:05:10 globals[1000]: <CRITICAL> (MainThread) Error trying to launch tor: [Errno 9] Bad file descriptor
```
A point of curiosity is how it says "sbws 0+unknown ...`, while on the real network, sbws would say "sbws 1.9.0 ..."

Then there's the error about bad file descriptor. I grepped in the codebase of sbws, and found:
```py
    torrc = parse_user_torrc_config(torrc, conf["tor"]["extra_lines"])    
    # Finally launch Tor    
    try:    
        # If there is already a tor process running with the same control    
        # socket, this will exit here.    
        stem.process.launch_tor_with_config(     
            torrc, init_msg_handler=log.debug, take_ownership=True    
        )    
    except Exception as e:    
        fail_hard("Error trying to launch tor: %s", e)    
    log.info("Started own tor.") 
```
So it seems like something went wrong when launching Tor with stem. I haven't examined stem's source code yet, but the comment about another Tor process already running seemed to be a clue.

So I tried to put sbws onto another virtual host. To ensure it's really on a different virtual host, I added a new node in the network graph file (gml). I thought I only need to connect the new node to an existing node to make it fully **connected**, as required by Shadow [here](https://shadow.github.io/docs/guide/network_graph_overview.html). But it complained about nodes not being directly connected, so I used NetworkX to add an edge from the new node to all other nodes. I do find this a little curious, as well as how the IPs of nodes in the gml and the IPs of virtual hosts in `shadow.config.yaml` are completely different, but maybe that's not so relevant right now.

Here's the virtual host entry in my `shadow.config.yaml` under the `tornet-0.01/` directory, manually modified after it's generated by tornettools.
```
  bwauthority:    
    network_node_id: 2521    
    ip_addr: 100.0.0.100     
    bandwidth_down: 1000000 kilobit    
    bandwidth_up: 1000000 kilobit      
    processes:    
    - path: /home/user/venv/bin/python3    
      args: /home/user/venv/bin/sbws -c /home/user/.sbws.ini scanner       
      environment: { PATH: "/home/user/venv/bin:/home/user/.local/bin:/home/user/.nvm/versions/node/v21.6.2/bin:/home/user/.cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/opt/puppetlabs/bin:/home/user/tor/src/core/or:/home/user/tor/src/app:/home/user/tor/src/tools" }
      start_time: 310    
      expected_final_state: running    
    - path: /home/user/venv/bin/python3    
      args: /home/user/venv/bin/sbws -c /home/user/.sbws.ini generate       
      environment: { PATH: "/home/user/venv/bin:/home/user/.local/bin:/home/user/.nvm/versions/node/v21.6.2/bin:/home/user/.cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/opt/puppetlabs/bin:/home/user/tor/src/core/or:/home/user/tor/src/app:/home/user/tor/src/tools" }
      start_time: 400    
```
The `network_node_id` matches the new node in the gml. The `ip_addr` is a new one, though sharing the same `100.0.0.0/24` subnet with 4uthority1, 2, 3, etc. New virtual hosts are also added for the simple server.

However, even after a new node and a new virtual host is configured, sbws in Shadow still threw the same error message. I'm wondering if tornettools or Shadow automatically start up some Tor-related stuff that conflict with sbws? If not, do you have any clues about what caused it?

Thanks again.

**Reply by sporksmith** [reply]
> I have tried both apache and nginx, but the former had some trouble with systemd, bus, and sudo, while the latter ran into some unexpected EOF when being requested with curl in Shadow.

I don't think we've tried apache, but we do have some notes for using nginx. https://shadow.github.io/docs/guide/compatibility_notes.html?highlight=nginx#nginx

But your workaround of using a simple python server instead is fine afaik

> A point of curiosity is how it says "sbws 0+unknown ...`, while on the real network, sbws would say "sbws 1.9.0 ..."

I wonder if it could be starting a different sbws instance than you intended? Processes in shadow don't inherit shadow's environment, including PATH, which can sometimes be a little surprising. OTOH I see you're already explicitly setting the PATH. Maybe in a shell try setting your PATH to the one you're using in the simulation and verify what `sbws` gets used?

> So it seems like something went wrong when launching Tor with stem. I haven't examined stem's source code yet, but the comment about another Tor process already running seemed to be a clue.
...
> Here's the virtual host entry in my shadow.config.yaml under the tornet-0.01/ directory, manually modified after it's generated by tornettools.

Is it the `sbws generate` process that's failing? It looks like the `sbws scanner` command would indeed still be running? (I'd imagine probably not though; that the generate command is probably intended to work this way...)

shadow's log (especially any warnings) would be helpful to debug further. an strace log for the failing process would also be helpful. https://shadow.github.io/docs/guide/shadow_config_spec.html?highlight=strace#experimentalstrace_logging_mode

**Reply by robgjansen** [reply]
You have two different processes using the same `/home/user/.sbws.ini` file, which is a red flag for me. Are those two processes configured to do something that conflicts with each other? I don't know much about sbws, but pasting your `.swbs.ini` file might help me guess.

The other red flag for me is the use of stem. You need to use stem in a specific way to make sure it is using your provided (ie simulated) tor process rather than trying to spin up its own new Tor process outside of the simulation. The code you posted (`stem.process.launch_tor_with_config`) does make it seem like it is launching tor itself, which hopefully would get launched _within_ the simulation environment, but I think it would help to see the config that is being provided as input to stem. It may be possible that the two different stems are both launching tor in a conflicting way (eg, if both are trying to launch a tor that binds to the same ports, the second one would probably fail).

And yeah, the strace logs from the `bwauthority` processes would help too.

**Reply by higher-order** [reply]
Thanks for your responses @sporksmith @robgjansen. Let me address those potential causes in separate comments.

From my understanding, sbws is designed so that `sbws scanner` will keep running and doing the actual measuring of the relays, while `sbws generate` should be invoked periodically to generate the v3bw file based on the measurement. It seems like `sbws scanner` will keep writing intermediate files under `~/.sbws/`, and `sbws generate`, when invoked, reads from there to generate the v3bw file.

From sbws's docs [here](https://tpo.pages.torproject.net/network-health/sbws/how_works.html), it says:
> The scanner measures the bandwidth of each relay in the Tor network (except the directory authorities) by creating a two hops circuit with the relay. It then measures the bandwidth by downloading data from a destination Web Server and stores the measurements.
The generator read the measurements, aggregates, filters and scales them using torflow’s scaling method.
Then it generates a bandwidth list file that is read by a directory authority to report relays’ bandwidth in its vote.

So `sbws scanner` will launch Tor, while `sbws generate` will not. That is why I assume that I can have one `sbws scanner` and multiple `sbws generate` run in a single virtual host in Shadow, at different times.

Here is the `~/.sbws.ini`:
```ini
[scanner]    
country = US    
dirauth_nickname = tor26    
    
[destinations]    
dest1 = on    
    
[destinations.dest1]    
url = https://3.3.3.3/1G    
verify = False    
country = US    
    
#[tor]    
#external_control_port = 57050 
```
I'm now using different IPs than `100.0.0.0/24`. But otherwise it's pretty simple.
Therefore I think having `sbws scanner` and `sbws generate` reading the same `~/.sbws.ini` isn't the problem here.

**Reply by higher-order** [reply]
I was concerned that Shadow didn't have the same environment too. IIRC I installed sbws from source under a virtual environment, though I may have remembered incorrectly. Either way, I deactivated the virtual environment, did `export PATH=` followed by the string I had in `shadow.config.yaml`, and ran `/home/user/venv/bin/python3 /home/user/venv/bin/sbws -c /home/user/.sbws.ini scanner`. Even without the virtual environment, the scanner could run and the generator could generate the v3bw files. It worked just as well with the virtual environment too.

**Reply by higher-order** [reply]
It was the scanner process that failed, rather than the generator. Under `shadow.data/hosts/bwauthority/`, there are `python3.10.1000.*` and `python3.10.1001.*`. I checked `python3.10.1001.stdout` and it had "No recent results, so not generating anything. (Have you ran sbws scanner recently?)". So 1001 is for the generator. And during the simulation, Shadow output "process 'bwauthority.python3.10.1000' exited with status Normal(1); expected end state was running but was exited: 1", which indicated that the scanner is supposed to keep running, but exited (with the bad file descriptor error).

I've attached `shadow.log` and the strace log of the failed scanner process here. Compression and suffix are applied for upload.
[python3.10.1000.strace.txt](https://github.com/shadow/shadow/files/14819000/python3.10.1000.strace.txt)
[shadow.log.zip](https://github.com/shadow/shadow/files/14819001/shadow.log.zip)

**Reply by higher-order** [reply]
> The other red flag for me is the use of stem. You need to use stem in a specific way to make sure it is using your provided (ie simulated) tor process rather than trying to spin up its own new Tor process outside of the simulation... I think it would help to see the config that is being provided as input to stem.

I wasn't aware that things like stem can run outside of Shadow. Even though the host machine isn't running any Tor, and the generator shouldn't be launching Tor, I guess then it's possible for the scanner to use stem to launch Tor outside of the simulation, which then conflicts with the Tor processes run by Shadow.

sbws generates the torrc config dynamically in its code and passes it to stem. Some of the code is straightforward, like how it uses a `TORRC_STARTING_POINT` and then does dictionary updates over it. But it will take me some more efforts to fully understand it. Instead of further examining the code, I added:
```py
for k, v in torrc.items():
  print(k, v)
```
To print out the torrc right after it's finalized and right before it's passed to stem. It prints:
```
conf <configparser.ConfigParser object at 0x7ffff4dca0b0>
SocksPort auto
CookieAuthentication 1
UseEntryGuards 0
UseMicrodescriptors 0
SafeLogging 0
LogTimeGranularity 1
ProtocolWarnings 1
FetchDirInfoEarly 1
FetchDirInfoExtraEarly 1
FetchUselessDescriptors 1
LearnCircuitBuildTimeout 0
DataDirectory /home/user/.sbws/tor
PidFile /home/user/.sbws/tor/tor.pid
ControlSocket /home/user/.sbws/tor/control
Log ['NOTICE file /home/user/.sbws/tor/log/notice.log']
CircuitBuildTimeout 60
```
Relevant code starts from [here](https://gitlab.torproject.org/tpo/network-health/sbws/-/blob/m19/sbws/util/stem.py?ref_type=heads#L289)

Do you see anything that can cause stem to not work well with Shadow?

If you want to examine the codebase, it starts from this line [here](https://gitlab.torproject.org/tpo/network-health/sbws/-/blob/m19/sbws/util/stem.py?ref_type=heads#L289).

**Reply by sporksmith** [reply]
Near the end of the strace log we have:     

```
00:05:30.000000000 [tid 1000] pipe2([-166775296, 32767] (0x7fffffffd358), O_CLOEXEC) = 0
00:05:30.000000000 [tid 1000] pipe2([11, 12] (0x7fffffffd358), O_CLOEXEC) = 0
00:05:30.000000000 [tid 1000] pipe2([13, 14] (0x7fffffffd358), O_CLOEXEC) = 0
00:05:30.000000000 [tid 1000] fstat() = -9 (EBADF)
```

Given the error message "Jan 01 00:05:10 globals[1000]: <CRITICAL> (MainThread) Error trying to launch tor: [Errno 9] Bad file descriptor", I think this is what's causing it to bail out.

We don't have strace parameter logging implemented for this syscall, but in the shadow log at the same simulated timestamp we have:

```
00:03:29.140400 [1523969:shadow-worker] 00:05:30.000000000 [WARN] [bwauthority:2.2.2.2] [thread.rs:820] [shadow_rs::host::thread::export] A descriptor exists for fd=3, but it is not a legacy file. Returning NULL.
```

It seems likely that it's trying to call `fstat` on a pipe file descriptor that it just created, and that shadow doesn't support this syscall on
 pipe file descriptors yet. @stevenengler probably has a better idea of what it'd take to get this working.

## Other bits, probably not the immediate problem

### Accessing files outside the sim dir

I also noticed that it's accessing some files outside of the simulation directory:              

```
00:05:30.000000000 [tid 1000] openat(-100, "/etc/passwd", O_CLOEXEC, Mode(0x0)) = 7
...
00:05:30.000000000 [tid 1000] open("/home/user/.sbws/state.dat.lockfile", O_RDWR | O_CREAT | O_TRUNC | O_CLOEXEC, Mode(S_IRWXU | S_IRWXG | S_IRWXO)) = 7
...
00:05:30.000000000 [tid 1000] open("/home/user/.sbws/state.dat", O_CLOEXEC, Mode(S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH)) = 9
```

This *might* not actually cause problems if there's only one sbws scanner in the simulation, and none running on the host machine. It looks like it's trying to get the current user's home directory by reading `/etc/passwd`, which might be a little annoying to work around. It's possible though that this is a fallback path, and that setting the `HOME` env variable might cause it to use that instead. As a test you could just set it to some hard coded absolute path, but a nicer solution might be to set it to something inside the host's simulation directory using a wrapper shell script something like `export HOME="$(pwd)/home"; mkdir "$HOME"; python3 bws`

It also looks like it's doing some file system locking, presumably on the lockfile. shadow's implementation of this is fake - the lock operation will always succeed, even if something else in the simulation has also locked that file. I don't think this is likely to be the problem, though.

### alarm syscall

There's also a failed `alarm` syscall near the end of the strace log. Given the error message that it's showing before failing, I don't think this is the cause of the failure, but might be another problem once we get past the `fstat` issue. We haven't implemented the `alarm` syscall yet, but from a quick look I think it'd be easy to do so. **edit**: implemented in #3321

**Reply by stevenengler** [reply]
> It seems likely that it's trying to call `fstat` on a pipe file descriptor that it just created, and that shadow doesn't support this syscall on pipe file descriptors yet. @stevenengler probably has a better idea of what it'd take to get this working.

I thought we ran into this before, but I couldn't find any issues or discussions. But yeah, it's likely calling `fstat` on the pipe:

```bash
strace python3 -c 'import subprocess; subprocess.run(["ls", "-l", "/dev/null"], capture_output=True)'
```

```strace
pipe2([3, 4], O_CLOEXEC)                = 0
pipe2([5, 6], O_CLOEXEC)                = 0
fstat(3, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
ioctl(3, TCGETS, 0x7fff5d4ffa70)        = -1 ENOTTY (Inappropriate ioctl for device)
lseek(3, 0, SEEK_CUR)                   = -1 ESPIPE (Illegal seek)
fstat(5, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
ioctl(5, TCGETS, 0x7fff5d4ffa70)        = -1 ENOTTY (Inappropriate ioctl for device)
lseek(5, 0, SEEK_CUR)                   = -1 ESPIPE (Illegal seek)
```

Implementing `fstat` for pipe fds should be straightforward and the calling code is probably only doing a `S_ISFIFO(stat.st_mode)`, but we would also need to fill in the rest of the `struct stat`. I'm not sure what we would put in `st_ino`, `st_dev`, etc, but probably any dummy values would be good enough. For things like `st_atime` we would probably need to check what linux returns for those since I'm not sure if access/modified/etc times are supported for pipes. I think `st_size` does return the pipe's size (probably amount of data buffered?) on linux, but doesn't on freebsd for example, so posix probably doesn't specify what happens here.

I'll open an issue for this (**edit:** #3319). @higher-order if you want to give it a shot, let us know and we can help point you in the right direction.

From the above strace, it looks likely that python will also later call `lseek` on the pipe fd, so that might need to be added too if python is checking the exact errno value. It would be easy though since it should always return `ESPIPE` for pipes, whereas shadow currently returns `EBADF`. (**edit:** fixed in #3320)

**Reply by sporksmith** [reply]
You *might* be able to work around this issue by starting a `tor` process directly from the shadow config, and configuring `sbws` to use that. See https://tpo.pages.torproject.net/network-health/sbws/config_tor.html and the `tor` section of https://tpo.pages.torproject.net/network-health/sbws/config.html

**Reply by higher-order** [reply]
Thanks for looking into this @sporksmith @stevenengler. I don't know Linux well enough to understand the innerworking of Shadow yet. But if I do decide to implement the syscalls myself later on, I'll probably start with reading`docs/contributing.md` and the developer documentation, as well as understanding those syscalls. Let me know if there are other resources I shouldn't miss.

The method of starting Tor separately and letting sbws connect to it worked! I copied the Tor process from a `perfclient10exit`, with an additional param for the control port. Here's the additional Tor process on the `bwauthoriy` virtual host, alongside its sbws scanner process and generator process. The `torrc` and `torrc-defaults` are also copied from `perfclient10exit`.
```
    - path: ~/.local/bin/tor
      args: --Address relaysbws --Nickname relaysbws --defaults-torrc torrc-defaults -f torrc --ControlPort 57050
      environment:
        OPENBLAS_NUM_THREADS: '1'
      start_time: 240
      expected_final_state: running
```
I added those lines to the config of sbws:
```
[tor]    
external_control_port = 57050 
```
sbws got past the bad file descriptor error and at least started trying to measure. From `python3.10.1003.stdout`:
```
Jan 01 00:05:30 sbws[1003]: <INFO> (MainThread) sbws 0+unknown with python 3.10.12 on shadowsys-shadowrelease-shadowmachine-64bit, stem 1.8.2, and requests 2.31.0
conf <configparser.ConfigParser object at 0x7ffff4dc5e70>
Jan 01 00:05:30 stem[1003]: <INFO> (MainThread) Connected to tor via port 57050
Jan 01 00:05:30 stem[1003]: <ERROR> (MainThread) Incorrectly configured SocksPort, should be auto, is 9050
NoneType: None
Jan 01 00:05:30 stem[1003]: <ERROR> (MainThread) Incorrectly configured CookieAuthentication, should be 1, is 0
NoneType: None
Jan 01 00:05:30 stem[1003]: <ERROR> (MainThread) Incorrectly configured UseMicrodescriptors, should be 0, is auto
NoneType: None
Jan 01 00:05:30 stem[1003]: <ERROR> (MainThread) Incorrectly configured FetchDirInfoEarly, should be 1, is 0
NoneType: None
Jan 01 00:05:30 stem[1003]: <ERROR> (MainThread) Incorrectly configured FetchDirInfoExtraEarly, should be 1, is 0
NoneType: None
Jan 01 00:05:30 stem[1003]: <ERROR> (MainThread) Incorrectly configured FetchUselessDescriptors, should be 1, is 0
NoneType: None
Jan 01 00:05:30 stem[1003]: <ERROR> (MainThread) Incorrectly configured LearnCircuitBuildTimeout, should be 0, is 1
NoneType: None
Jan 01 00:05:30 stem[1003]: <INFO> (MainThread) Started or connected to Tor 0.4.9.0-alpha-dev (git-df503f464efcbfaf).
Jan 01 00:05:30 relaylist[1003]: <INFO> (MainThread) Number of consensuses obtained in the last 0 days: 1.
Jan 01 00:05:30 resultdump[1003]: <INFO> (Thread-1 (enter)) Reading and processing previous measurements.
Jan 01 00:05:30 scanner[1003]: <INFO> (MainThread) Started the main loop to measure the relays.
Jan 01 00:05:30 scanner[1003]: <INFO> (MainThread) In the executor, queue all future measurements.
Jan 01 00:05:30 relayprioritizer[1003]: <INFO> (MainThread) Spent 0.082016 msecs calculating relay best priority
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_0) The consensus implements congestion control.
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_1) The consensus implements congestion control.
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_2) The consensus implements congestion control.
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_0) The consensus says to use exits that support congestion control.
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_1) The consensus says to use exits that support congestion control.
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_2) The consensus says to use exits that support congestion control.
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_0) The consensus implements congestion control.
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_1) The consensus implements congestion control.
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_2) The consensus implements congestion control.
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_0) The consensus says to use exits that support congestion control.
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_1) The consensus says to use exits that support congestion control.
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_2) The consensus says to use exits that support congestion control.
Jan 01 00:05:30 scanner[1003]: <INFO> (MainThread) Measurement ready: [<sbws.lib.resultdump.ResultErrorSecondRelay object at 0x7ffff454abf0>]
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_0) The consensus implements congestion control.
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_1) The consensus implements congestion control.
Jan 01 00:05:30 relaylist[1003]: <INFO> (measurer_2) The consensus implements congestion control.
Jan 01 00:05:30 scanner[1003]: <INFO> (MainThread) Pending measurements: 49 out of 50:
...
```
Even though it reported incorrect Tor configs, I checked the code of sbws and saw that it automatically applies its own settings in such cases.

However, at the last few lines shown here, it reported "Measurement ready" with `ResultErrorSecondRelay`.
Some comments on this error class:
```py
class ResultErrorSecondRelay(ResultError):    
    """    
    Error when it could not be found a second relay suitable to measure    
    a relay.    
    
    A second suitable relay is a relay that:    
    - Has at least equal bandwidth as the relay to measure.    
    - If the relay to measure is not an exit, \    
      the second relay is an exit without `bad` flag and can exit to port 443.    
    - If the relay to measure is an exit, the second relay is not an exit.    
    
    It is instantiated in :func:`~sbws.core.scanner.measure_relay`.    
    
    .. note:: this duplicates code and add more tech-debt,    
       since it's the same as the other    
       :class:`~sbws.lib.resultdump.ResultError` classes except for the    
       ``type``.    
       In a future refactor, there should be only one ``ResultError`` class    
       and assign the type in the ``scanner`` module.    
    """ 
```
Some comments about the `error_no_helper` function, which seems to be the only (?) function that constructs `ResultErrorSecondRelay`:
```py
    # When `error_no_helper` is triggered because a helper is not found, what
    # can happen in test networks with very few relays, it returns a list with
    # the error.
```
So it seems possible that measurements result in `ResultErrorSecondRelay` because the 1% Tor net I'm simulating with Shadow is not big enough. However, it still gave the same result when I simulated a 10% Tor net. And this error happens with measurements of all relays (omitted from my quote of `python3.10.1003.stdout`, not just a few of them, which I find a little suspicious.

Right now I'm waiting for a more powerful machine to simulate bigger networks, as a 20% Tor net exhausts the RAM and will take a long time to finish. But do you know anything in sbws or Shadow that can cause **all** measurement to fail with `ResultErrorSecondRelay`, other than the network being too small?

**Reply by sporksmith** [reply]
Glad that helped!

In your position my next move would probably be to add some debug logging in the sbws code to find for which reason it's rejecting each relay.
